### ISSUE TITLE
MakeCode 2022 consumes more memory and some v4 programmes won't download anymore

### ISSUE THREAD

**Issue by microbit-carlos** [microsoft/pxt-microbit#4788]
**Describe the bug**
We've got several reports from users where their micro:bit MakeCode projects stopped working after MakeCode 2022 release date.

**To Reproduce**
1. Download this hex file: [microbit-fits-v4_does-not-v5.hex.zip](https://github.com/microsoft/pxt-microbit/files/9028251/microbit-fits-v4_does-not-v5.hex.zip)
2. Load it in https://makecode.microbit.org
3. Click "Download"
4. The error `program too big by 6184 bytes!` appears

The same hex file in https://makecode.microbit.org/v4 downloads correctly.

Rather than share one of the user programmes, I create this hex file by adding a bunch of extensions and just randomly dropping blocks on the workspace until it stopped working on MakeCode v5 and worked in v4. So the programme in itself will not make much sense.

**Expected behavior**

Existing programmes to still work.

**Screenshots**

<img width="317" alt="image" src="https://user-images.githubusercontent.com/29712657/176880567-72da8d50-1357-4f1d-8a12-4a19f431b2e6.png">

**micro:bit version (please complete the following information):**

N/A

**Desktop (please complete the following information):**
 - OS: macOS 12
 - Browser: Chrome
 - Version: 103

**Smartphone (please complete the following information):**

**Additional context**

makecode.microbit.org version:  [5.0.6](https://github.com/Microsoft/pxt-microbit/releases/tag/v5.0.6)
Microsoft MakeCode version:  [8.0.3](https://github.com/Microsoft/pxt/releases/tag/v8.0.3)
microbit runtime version:  [v2.2.0-rc6](https://github.com/lancaster-university/microbit/releases/tag/v2.2.0-rc6)
codal-microbit-v2 runtime version:  [v0.2.40](https://github.com/lancaster-university/codal-microbit-v2/releases/tag/v0.2.40)

**Comment by microbit-carlos** [comment]
With the provided hex file in the MakeCode 2022 editor, if I then add the Data Logger extension, which is configured to be V2 only, then MakeCode successfully compiles/downloads the programme.

So the memory issue might right now mostly affect V1 (or at least more easily in V1), and as V1 DAL hasn't been updated vs MakeCode v4, it suggest all the additional memory comes from pxt updates.

**Comment by jaustin** [comment]
@abchatra @mmoskal we've had more reports of this than anything else and it's a regression for these users, so would like to see if we can investigate. Especially because it affects V2 users even though the constraint appears to be related to V1, really?

**Comment by abchatra** [comment]
@microbit-carlos @jaustin is it possible to create a repro without the extension? If you can do that, I would to figure out the exact checkin.

**Comment by microbit-carlos** [comment]
It really is just copy and paste code until just about you reach the limit in V4 and then load the hex in MakeCode 2022:
[microbit-v4large.hex.zip](https://github.com/microsoft/pxt-microbit/files/9044808/microbit-v4large.hex.zip)

**Comment by microbit-carlos** [comment]
While this issue has been focused on flash consumption, it looks like there is also a higher RAM consumption by MakeCode:
- https://github.com/microsoft/pxt-microbit/issues/4808

**Comment by riknoll** [comment]
it appears that this commit is the culprit: https://github.com/microsoft/pxt-common-packages/commit/9454016e1e34b0fc0cfc8e5c86a4fc1797637c49

**Comment by mmoskal** [comment]
Strange. The commit definitely looks like a noop.

**Comment by riknoll** [comment]
well, it appears to be more than just that. delta debugging between the two versions @abchatra mentioned led me to that commit and checking out the commit before does fix the issue. however, reverting that commit on the stable branch that pxt-microbit points too does not fix the issue.

i've narrowed it down to libs/base and libs/radio, but there are no obvious suspects in the diffs between the old code and the new. checking out the old libs/base and libs/radio on the stable branch does reduce the code size quite a bit though.

**Comment by mmoskal** [comment]
You could try to run this script https://github.com/microsoft/jacdac-stm32x0/blob/main/scripts/map-file-stats.js on the .map file before and after (requires local build)

**Comment by riknoll** [comment]
Thanks @mmoskal, I'll give it a shot!

**Comment by riknoll** [comment]
well it took me awhile to figure out how to get yotta to run! I ended up having to use Ubuntu 18.04.

Anyway, here are the things that changed between version 4.1.5 (left side of arrow) and 5.0.13 (right side of arrow)

```
RAM.libs/blocksprj/built/yt/source/radio/radio.cpp.o 1 -> 2
RAM.libs/blocksprj/built/yt/source/core/pins.cpp.o 10 -> 11
RAM.TOTAL 2692 -> 2694

DATA.libs/blocksprj/built/yt/source/pointers.cpp.o 1252 -> 1268
DATA.TOTAL 9098 -> 9114

TEXT.libs/blocksprj/built/yt/source/radio/radio.cpp.o 408 -> 508
TEXT.libs/blocksprj/built/yt/source/core/serial.cpp.o 544 -> 548
TEXT.libs/blocksprj/built/yt/source/core/codal.cpp.o 820 -> 818
TEXT.libs/blocksprj/built/yt/source/core/pins.cpp.o 1064 -> 1108
TEXT.ym/microbit-dal/source/microbit-dal.a 34882 -> 34996
TEXT.TOTAL 114738 -> 114968

FLASH.libs/blocksprj/built/yt/source/radio/radio.cpp.o 408 -> 508
FLASH.libs/blocksprj/built/yt/source/core/serial.cpp.o 544 -> 548
FLASH.libs/blocksprj/built/yt/source/core/codal.cpp.o 836 -> 834
FLASH.libs/blocksprj/built/yt/source/core/pins.cpp.o 1149 -> 1193
FLASH.libs/blocksprj/built/yt/source/pointers.cpp.o 1278 -> 1294
FLASH.ym/microbit-dal/source/microbit-dal.a 37391 -> 37475
FLASH.TOTAL 123836 -> 124082
```

**Comment by pelikhan** [comment]
radio.on , radio.off ?

**Comment by pelikhan** [comment]
This script is great! We should run it on every build :)

**Comment by riknoll** [comment]
I tried reverting every file it listed as changed back to 4.1.5 and I'm still getting a "program too big by 5216 bytes". @mmoskal any pointers on how to interpret these results?

**Comment by mmoskal** [comment]
It could be that the size of compiled code increased - you could look at the assembly output in the editor - it has comments on top that specify sizes. To view the .asm file I think you have to disable the simulator otherwise it disappears very quickly. You could also diff the .asm but you may need to replace some/all numbers by a placeholder to remove "random" changes due to different AST node IDs etc.

I guess compiled code increase is more likely here - the 200 byte change above shouldn't be that big of a problem.

**Comment by AlasdairAtKitronik** [comment]
@microbit-carlos @abchatra Any update on this is it's definitely still an issue?

**Comment by abchatra** [comment]
Thanks for the reminder @AlasdairAtKitronik. I will update by EOW on the progress.

**Comment by abchatra** [comment]
@AlasdairAtKitronik makecode.microbit.org/stable should have a workaround with letting users to download on v2. Please take a look.

**Comment by abchatra** [comment]
https://github.com/microsoft/pxt-microbit/pull/4824#issuecomment-1220953203

**Comment by abchatra** [comment]
We will release after some testing.

**Comment by AlasdairAtKitronik** [comment]
@abchatra I have tested the workaround and it works for downloading the program for V2 only.
I take it this is just a temporary fix while you work out how to reduce the download size so programs fit on V1 as well?

**Comment by abchatra** [comment]
Yotta package is the reason for hex being bigger in the latest release. Plan here is to move DAL (used by V1) to use CODAL as well, which gets rids of yotta package. This is going to take some time, meanwhile we will ship this workaround for V2.

@jaustin @finneyj - FYI

**Comment by riknoll** [comment]
@mmoskal

**Comment by mmoskal** [comment]
I think it was update of build environment to latest GCC, which was apparently somehow the only way to run latest yotta that was required due to yotta being killed by arm.

I'm not sure CODAL will use less flash than DAL, but you're the expert.

Also I my understanding is that the difference in flash usage wasn't large, it was just that the programs were already at the edge, is that right @riknoll  ?

**Comment by riknoll** [comment]
yes, it was pretty close already.

when we updated yotta, it forced us to update the version of ubuntu that our build service is using because the python libs no longer supported ssl on the older version. the gcc updates were incidental to that.

**Comment by microbit-carlos** [comment]
@riknoll could you point me to the old and new Dockerfile? Yotta hasn't been updated in years and I think we could find a combination of Ubuntu release and dependencies that can keep the same gcc version. Worse case scenario, if the issue is only the GCC version used, we can install whichever version we need by downloading a specific released from launchpad, as Arm GCC builds are hosted there and can be uncompressed and used directly without an apt package (which might be bound to an Ubuntu release).

**Comment by microbit-carlos** [comment]
Btw, I am a bit surprised that updating GCC resulted in a larger flash and RAM consumption, it's usually been the other way around. But if the fix is to get a docker image with the same GCC version, I think that should be doable.

In terms of numbers, we are talking about _at least_ 6 KBs of flash, which yes, it's a bit on the edge, but we get to that point easily by using couple of large extensions. On the other hand for RAM I'm not too sure how much extra memory is being consumed in V5, but it does surprise me that a newer version of GCC results in more RAM usage.

**Comment by riknoll** [comment]
For the record, I was just talking about flash size and not RAM. Haven't personally done any investigation into memory consumption.

You can find the dockerfile under pext/yotta in dockerhub:

https://hub.docker.com/r/pext/yotta/tags

We currently use the `update-yotta3` tag and were previously using the `gcc4` tag

**Comment by microbit-carlos** [comment]
Thanks @riknoll, do you have a link to the Dockerfile in GitHub, and a "before" or "after" commit? Something I can close locally and modify? I can't find a way to the `gcc4` dockerfile on DockerHub.

**Comment by riknoll** [comment]
ah, all the old actually does is extend the mbed/yotta dockerfile:

https://github.com/ARMmbed/yotta-docker

Can't link a commit because our backend isn't public

**Comment by jaustin** [comment]
> I'm not sure CODAL will use less flash than DAL, but you're the expert.

@mmoskal just to clarify this, I don't think using the whole of CODAL is on the table - just build.py and associated tooling. So, still building microbit-dal but using CODAL's toolchain to do that.

I don't think we think this will solve the problem in this issue, but will at least make debugging it easier as we can build DAL without needing yotta and dependencies which have become a pain to install (and use). 

@JohnVidler I think yotta even has a switch for dumping the cmake it would use to build, so you may be able to smooth your path with that.

**Comment by microbit-carlos** [comment]
Running a quick test building DAL (from the https://github.com/lancaster-university/microbit-samples repo) with the old PXT Yotta docker image and the newest one, we can see a decrease in flash usage (at least from the DAL side).

Both examples were built using the following command:
```
$ docker run -v $(pwd):/home --rm pext/yotta:<label_here> bash -c "cd /home && yt clean && yt build"`
```

### pext/yotta:gcc4
```
$ docker run -it --rm pext/yotta:gcc4 arm-none-eabi-gcc --version
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.2.1 20151202 (release) [ARM/embedded-5-branch revision 231848]
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
$ arm-none-eabi-size build/bbc-microbit-classic-gcc/source/microbit-samples
   text    data     bss     dec     hex filename
  67968     168    2032   70168   11218 build/bbc-microbit-classic-gcc/source/microbit-samples
```
```
$ arm-none-eabi-size build/bbc-microbit-classic-gcc/source/microbit-samples --format=sysv
build/bbc-microbit-classic-gcc/source/microbit-samples  :
section              size        addr
.text               67960       98304
.ARM.exidx              8      166264
.data                 168   536879104
.bss                 2032   536879272
.heap                2048   536881304
.stack_dummy         2048   536881304
.ARM.attributes        40           0
.comment              110           0
.debug_info       1352290           0
.debug_abbrev      118267           0
.debug_loc         136664           0
.debug_aranges      12232           0
.debug_ranges       22376           0
.debug_line        130827           0
.debug_str         173591           0
.debug_frame        34876           0
.stab                  60           0
.stabstr              118           0
Total             2055715
```


### pext/yotta:update-yotta3

```
$ docker run -it --rm pext/yotta:update-yotta3 arm-none-eabi-gcc --version
arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2018-q3-update) 7.3.1 20180622 (release) [ARM/embedded-7-branch revision 261907]
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
$ arm-none-eabi-size build/bbc-microbit-classic-gcc/source/microbit-samples
   text    data     bss     dec     hex filename
  62048     168    2032   64248    faf8 build/bbc-microbit-classic-gcc/source/microbit-samples
```
```
$ arm-none-eabi-size build/bbc-microbit-classic-gcc/source/microbit-samples --format=sysv
build/bbc-microbit-classic-gcc/source/microbit-samples  :
section              size        addr
.text               62040       98304
.ARM.exidx              8      160344
.data                 168   536879104
.bss                 2032   536879272
.heap                2048   536881304
.stack_dummy         2048   536881304
.ARM.attributes        40           0
.comment              127           0
.debug_info       1456307           0
.debug_abbrev      132477           0
.debug_loc         140908           0
.debug_aranges      11736           0
.debug_ranges       24240           0
.debug_line        139841           0
.debug_str         175770           0
.debug_frame        33320           0
.stab                  60           0
.stabstr              118           0
Total             2183288
```

### Comparison

```
  70168  -  64248   = 5920
(gcc v5)   (gcc v7)
```

So there has been a reduction in flash consumption from DAL code with the latest yotta docker image using GCC v7.
It's possible that the rest of the pxt C++ code might be going the opposite direction, but I'd be surprised.

### Adding GCC v5 to a newer Docker image

If you want to create a new Docker image with a newer Ubuntu and GCC v5.2.1 you can replace the installation from the `team-gcc-arm-embedded` ppa with something like this:

```Dockerfile
WORKDIR /opt/
RUN cd /opt/ && \
    wget -q https://launchpad.net/gcc-arm-embedded/5.0/5-2015-q4-major/+download/gcc-arm-none-eabi-5_2-2015q4-20151219-linux.tar.bz2 && \
    echo "f88caac80b4444a17344f57ccb760b90  gcc-arm-none-eabi-5_2-2015q4-20151219-linux.tar.bz2" >> MD5SUM && \
    md5sum -c MD5SUM && \
    rm MD5SUM && \
    tar -xf gcc-arm-none-eabi-5_2-2015q4-20151219-linux.tar.bz2 && \
    rm gcc-arm-none-eabi-5_2-2015q4-20151219-linux.tar.bz2
ENV PATH $PATH:/opt/gcc-arm-none-eabi-5_2-2015q4/bin
```

**Comment by abchatra** [comment]
@microbit-carlos can you summarize the feedback? Are you saying we should move to the latest GCC?

**Comment by microbit-carlos** [comment]
The summary is that the old Yotta docker image had GCC v5, and the latest one has GCC v7. It is my understanding from this thread, that we believe that moving from GCC v5 to v7 has caused the additional flash consumption, but compiling a DAL compilation with those two docker images shows that GCC v7 (the one PXT uses now) actually reduces flash consumption by almost 5 KBs.

Of course, MakeCode has more C/C++ code (does it also change other DAL or compiler settings? I don't know), and maybe compiling that with GCC v7 (compared with v5) does increases the flash consumption, I'd be surprised, but it is definitely possible. However I don't know how to test that, so my last comment also provided a way to add GCC v5 to your latest Dockerfile (which is also not public, so I cannot modify it myself), if you'd like to test it.

**Comment by abchatra** [comment]
Our docker-image file here: 
[docker-image.txt](https://github.com/microsoft/pxt-microbit/files/9889805/docker-image.txt)

**Comment by microbit-carlos** [comment]
Actually, the dockerhub `pext/yotta:latest` points to this image:
<img width="1288" alt="image" src="https://user-images.githubusercontent.com/29712657/198690654-ea48a874-6491-4afc-a358-771b3844defd.png">

Which was pushed 5 years ago, and uses GCC 6.2.1:

```
$ docker run -it --rm pext/yotta:latest  arm-none-eabi-gcc --version
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 6.2.1 20161205 (release) [ARM/embedded-6-branch revision 243739]
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

@abchatra @riknoll From this comment Makecode is currently using `pext/yotta:update-yotta3`, is that configured in a different repo?

> We currently use the `update-yotta3` tag and were previously using the `gcc4` tag

**Comment by riknoll** [comment]
yes, we don't touch the latest tag for historical reasons. it's still referenced in a few places but not for the micro:bit build

**Comment by microbit-carlos** [comment]
Right, so if `latest` tag is never touched, the pxt-microbit repo `pxtarget.json` file points to the `latest` tag, but the cloud builds are done with the `yotta3` tag, then I assume the cloud build tags are configured somewhere else?

Why not change ` pxt-microbit/pxtarget.json` to point to `yotta3` as well?

**Comment by mark2018888** [comment]
@jaustin Hello Jonathan, it looks like the problem of the "Program too big by 1420bytes" still exists in makecode.microbit.org version:  [5.0.12](https://github.com/Microsoft/pxt-microbit/releases/tag/v5.0.12), what we are doing now is to get back to the former 4.0+ version to make things work, but it could be complicated for Ss to operate and understand in classrooms, is there any possibility to instruct the due time to have it worked out in the up-to-date MakeCode? Appreciate it very much for your help:-).

**Comment by jaustin** [comment]
Currently in either beta or stable (https://makecode.microbit.org/stable#) this issue is resolved and you can download a hex file.

However @abchatra it would be great for web USB to work properly here, not just the hex download. In fact, when web USB is connected, I think we should just automatically flash the V2 device, surely?

**Comment by martinwork** [comment]
@jaustin When I try with stable or beta, the example in https://github.com/microsoft/pxt-microbit/issues/4861 doesn't build for V1.

Should the V2 only build be a universal hex with error 9xx for V1?

**Comment by microbit-carlos** [comment]
Yes, producing an Intel Hex file that can be drag&dropped into a V1 without showing some obvious error might be confusing as well.
I think @martinwork's idea to do a Universal Hex with an error code for the V1 section is much better.

**Comment by jaustin** [comment]
Hi @abchatra I'd really like to prioritise getting a better fix for this into live as we're seeing a few more people with this issue.

There are two things going on
1) Finding out why the size has increased so significantly. We've reached a point where we don't think we can help more, because when we build with newer toolchains the flash size decreases. We haven't got a local way to reproduce the MS setup where the code size increases. I'd like to continue to investigate this, but I think we need to be realistic that it won't be an instant answer. Who from MakeCode can sync up with @microbit-carlos to try and bottom this out?

2) Using /stable right now there aren't ideal mitigations in place, specifically
2a) the hex file you get is an intel hex, not a Universal hex, so can fail silently on V1 with no error code. Given we know the program won't fit on V1, we should show Error 92x (ideally a new error, but 927 would suffice if it's a blocker to do something else) for this build - as if we had a V2 only extension added.
2b) the mechanism that we use for this stops web USB from working, it's USB download only, which doesn't work on macOS Ventura and is also cumbersome

# So, I propose the following actions:

ASAP for users: Update /stable and /beta to:
* Generate universal hex in this situation, but where the V1 program just generates an error code (926 please)
* Allow web USB even after this error for V2 only (and show an error for V1 like we do for an incompatible extension. [1])

In the background:
* Debug the size increase with @microbit-carlos and someone from @abchatra's team - for which the next step is getting a local reproduction of the MakeCode build where we can inspect the binaries created and be sure about the toolchain versions at issue

For the next MakeCode release
* Deprecate yotta and build microbit-dal with build.py tools. @johnvidler is looking at this.


[1] like this ![image](https://user-images.githubusercontent.com/147113/206423643-f25934f6-50ea-4a0e-a0e6-da44b0b10433.png)
(If *absolutley* necessary to get things implemented sooner, I think it would be okay to show the "V2 only" message and then still push to a V1 over webUSB as long as the point above is done so they get an error on the display)

**Comment by abchatra** [comment]
My assumption was this is locally reproducible with this: [docker-image.txt](https://github.com/microsoft/pxt-microbit/files/9889805/docker-image.txt)
If you have tried this and the bug is not reproducible do let me know.

Lot of folks in my team are out of office. Tough to find someone to take this over till new year. I will try.

**Comment by jaustin** [comment]
> # So, I propose the following actions:
> ASAP for users: Update /stable and /beta to:
> 
> * Generate universal hex in this situation, but where the V1 program just generates an error code (926 please)
> * Allow web USB even after this error for V2 only (and show an error for V1 like we do for an incompatible extension. [1])

Hi @abchatra we're working through the 'in the background' bit, but not the ASAP bit. 

Is there anyone that could look at this please? The experience is quite confusing for someone who hits this problem....

Given #4934 the priority of resolving this aspect of things has increased for me.

@abchatra would you like me to file a new issue just for the Ux side and keep this one for the investigation on reasons for the size increase?

**Comment by abchatra** [comment]
Can I just release the fix in the makecode.microbit.org/stable? Blocking the user from downloading is bit too much.

**Comment by abchatra** [comment]
/stable basically warns and lets user download.

**Comment by jaustin** [comment]
There are two issues with /stable, though it's definitely better than live
* It doesn't let the user use web-usb
* it generates an intel hex, not a universal hex, which means it will fail silently/weirdly on a micro:bit V1. 

We really don't want V2 only non-universal hex files kicking around. Is it possible to generate a universal hex as if a V2-only extension had been added? (And the bonus is that I think this would also make web USB work)

**Comment by microbit-carlos** [comment]
Okay, from my side I've been able to do some progress, mostly thanks to @aznhassan `pxt` PR  and I am now able to build locally 🎉
- https://github.com/microsoft/pxt/pull/9363

There isn't a pxt release yet with the fix, so at the moment we need to clone the repo locally and link it to pxt-microbit.
@abchatra can a release be made soon? That'd save us a few CI steps we are adding to the CODAL repos to test with MakeCode.

I tried to create a MakeCode build with the `pext/yotta:latest` (old) vs `pext/yotta:update-yotta3` (new),
however there is a bug in the old version of Yotta (v0.16.4) in the `latest` docker image that means the build fails.
This wasn't a problem before the Yotta servers were shut down, as we were fetching the targets from there, but now that we are fetching them GitHub the bug is triggered and stops the build.

The last Yotta release v0.20.5 fixes the issue, so to be able to test in the old environment I had to create a new docker image based on `pext/yotta:latest` and update Yotta. This is the Dockerfile I've used to create a image to test `pext/yotta:latest`: https://gist.github.com/microbit-carlos/7069544b8d2d5ffc47dc5bd5a57c78f1

Okay, so finally with all that I was able to build MakeCode with different docker images, using the test programme at the top of the issue, and compare them.

- Original issue: program too big by 6184 bytes
- Current MakeCode deployment: program too big by 6196 bytes 
- Local `pext/yotta:update-yotta3`: program too big by 6256 bytes (similar, but not exactly the 
- Local `pext/yotta:latest` (modified with newer Yotta): program too big by 7280 bytes 


> **Note**

**So, as expected, the older yotta docker image with an older GCC version resulted in a larger build.**

------

These are the steps I used to build MakeCode locally.

First I made sure I don't have a global installation of the pxt cli tool:
```
sudo npm uninstall -g pxt pxt-core pxt-common-packages
```

```
git clone https://github.com/microsoft/pxt-microbit.git
git clone https://github.com/microsoft/pxt.git
cd pxt
npm install
npm run build
cd ../pxt-microbit
npm install
sudo npm link ../pxt
```

At this point we have to modifying the pxt-microbit/pxtarget.json file.

To change the docker image used we replace the two mentions of `pext/yotta:latest`
to whatever image we are testing.

For Yotta to be able to clone all the GitHub repositories we have to provide it
a GitHub Access Token. That was done by adding this environmental varaible to
the docker environment:

```json
    "variants": {
        "mbdal": {
            "compile": {},
            "compileService": {
                "dockerArgs": [
                    "--env",
                    "YOTTA_GITHUB_AUTHTOKEN=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
                ]
            }
        },
```

At this point we can build and serve:
```
npx pxt buildtarget --local
PXT_FORCE_LOCAL=1 PXT_RUNTIME_DEV=1 PXT_ASMDEBUG=1 npx pxt serve --local
```

**Comment by eanders-ms** [comment]
Interesting to note: The entrypoint for microbit v2 did not change between releases (all tests cases report `.startaddr 0x47000`).

**Comment by jaustin** [comment]
I think the question we really need to answer here, then, is what's the change in the toolchain?

As per https://github.com/microsoft/pxt-microbit/issues/4788#issuecomment-1426014425 when we build outside the MakeCode infrastructure (but with the MakeCode docker images) we see smaller builds from newer toolchains, not larger.

> current theory was that MakeCode v4 was using the pext/yotta:latest docker image, and MakeCode v5 is using pext/yotta:update-yotta3, with updated tooling versions, and this was causing the larger builds.

[... more results from @microbit-carlos's post]

> So pext/yotta:update-yotta3 produces **smaller** builds than pext/yotta:latest.

So is it that we've got the assumption wrong about which docker image is used when? Or has something else changed under the hood in the build system? Or perhaps looking at the number of bytes over is too crude to tell us what's going on? Perhaps updating the docker file/image with a newer yotta is also part of the story?

**Comment by eanders-ms** [comment]
Thinking in a different direction... can we leverage the fact that most microbit programs don't use bluetooth to give space back to the user? By excluding the soft device hex in these instances and adjusting FLASH start accordingly, that could potentially free up a lot of space for user code, to the tune of 80k if I'm not mistaken. I could easily be missing something here.

**Comment by abchatra** [comment]
@finneyj

**Comment by jaustin** [comment]
Hi @eanders-ms This is an appealing idea but for the foreseeable future with micro:bit avoiding SoftDevice is not possible.

The issue is that even if a program doesn't explicitly use Bluetooth, we rely on SoftDevice to be able to enter pairing and BLE update mode. This is required for people using iPads, tablets and phones. We are seeing an increasing number of schools use, or want to use, micro:bit with iPads and it's important that the experience is as close as possible there to everywhere else.

So, basically SoftDevice needs to be in the flash always, even if the user is not using BLE as part of their program, so that we can update the program over BLE.

@finneyj and co have been looking at alternative BLE stacks to provide this bootloader mode, but right now SoftDevice is the most flash efficient option we are aware of that has the properties we need. 

I think the primary areas I'm interested in for this are as in 
https://github.com/microsoft/pxt-microbit/issues/4788#issuecomment-1342490577

> ASAP for users: Update /stable and /beta to:
> 
> Generate universal hex in this situation, but where the V1 program just generates an error code (926 please)
> Allow web USB even after this error for V2 only (and show an error for V1 like we do for an incompatible extension. [1])
> In the background:
> 
> Debug the size increase with @microbit-carlos and someone from @abchatra's team - for which the next step is getting a local reproduction of the MakeCode build where we can inspect the binaries created and be sure about the toolchain versions at issue

I still don't understand why the trend of size that MakeCode team are seeing is opposite to the trend Carlos sees (newer builds give smaller binaries for Carlos).

**Comment by finneyj** [comment]
Thanks @jaustin @eanders-ms. 

Indeed, Softdevice is a hungry beast when it comes to FLASH and RAM memory... We do have a yotta target configured to run with Softdevice, but as Jonny said, that would leave a lot of our users stranded that rely of Bluetooth for programming...

Early indications from student studies indicate that we could get back upward of 40K of flash from switching BLE stack. This is not a trivial undertaking though - it'll take a lot of dev+test time to make sure we have a comparably good experience. BLE is a remarkably twitchy protocol, with a lot of edge cases where different stacks behave differently... There are also some security features that we currently enable that might not be available in the new stacks.

Although well worthy, until now it's not really been flagged as a high enough priority...

**Comment by eanders-ms** [comment]
OK, I confirmed that **GCC 5.4.1 + MakeCode 4.1.5 produces a test program that fits in memory** by a whopping 8 bytes!

The bad news is it still doesn't work for MakeCode v5. It is too big by 1204 bytes.

![image](https://github.com/microsoft/pxt-microbit/assets/12176807/e5f4c584-b2be-4894-bb5e-6f70d27d229a)

I noticed the start address (where user code is added to the hex) jumped by 1024 bytes in the v5 build. I take this to indicate the cpp additions were enough to push the start address out to the next page boundary.

**I think we have our culprits:**
* 5k loss due to GCC update
* 1k loss due to CPP additions

**Next steps:**
* Configure the build to use GCC 5 again
* Invalidate cached builds
* Discuss options for optimizing the cpp

**Comment by eanders-ms** [comment]
gcc5 toolchain for the microbit v1 build is live now. It is most noticeable in the beta channel, where you will get a new native hex built, while the release channel continues to use cached hexes (built with gcc7). gcc5 brings significant improvement to hex memory footprint, but doesn't get quite enough back for the test case here to work. Therefore keeping this bug open. We have additional work to do if we want to reduce the cpp footprint.

**Comment by microbit-carlos** [comment]
Thanks @eanders-ms, this is a great investigation on the memory consumption over different GCC toolchains!

Does that mean the current MakeCode live deployment builds V1 DAL with a docker image using gcc5, and V2 CODAL uses a docker image with gcc7?

If that's the case, could the https://github.com/microsoft/pxt-microbit/blob/master/pxtarget.json file be updated with these changes as well? Even if they are overwritten on the build server, it'd be really useful to have the pxtarget.json file to be as close as possible to the production configuration.  

If you still have the different builds around, specifically the ELF files (usually built as a `MICROBIT` file without an extension, they can be identified because they start with the magic number 0x7F454C46, which is 0x7 'E' 'L' 'F' ) as then  it'd be great to use something Bloaty to see where the extra flash is going:
https://github.com/google/bloaty

You can use this docker image directly, which might be quicker/simpler than building from source: https://github.com/carlosperate/bloaty-action/pkgs/container/bloaty#using-the-docker-image-to-run-bloaty-directly

**Comment by eanders-ms** [comment]
Oh yes, pxtarget.json should have been updated. Thanks for pointing that out. PR here: https://github.com/microsoft/pxt-microbit/pull/5265

Curiously, the V2 CODAL build uses gcc 6. I left that unchanged for the time being.

Bloaty looks very useful. Our backend compiler does not produce ELF files, though it we could change that if needed. In the course of troubleshooting this, I developed a small utility to help me understand how a hex file loads into memory: https://eanders-ms.github.io/mb-hex-explorer/. It was educational for me, if nothing else. Though it would be useful to integrate something similar into our ci build and be able to catch regressions like this earlier.

**Comment by microbit-carlos** [comment]
> Our backend compiler does not produce ELF files, though it we could change that if needed.

I thought it used the default CMake files from https://github.com/lancaster-university/microbit-v2-samples/ ? Are they being replaced, patched, or different ones used?

It'd be good for us to better understand how the backend compiler differs from the standard builds, as we have updated these in the past to accommodate some changes in the build system and/or configurations, and that might clash with what the backend compiler might be doing.

**Comment by thomasjball** [comment]
@abchatra - time to close this one?
